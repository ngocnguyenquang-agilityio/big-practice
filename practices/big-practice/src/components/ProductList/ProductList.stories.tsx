// Types
import type { Meta, StoryObj } from '@storybook/react';

// Components
import ProductList from '.';

const meta = {
  title: 'Components/ProductList',
  component: ProductList,
} satisfies Meta<typeof ProductList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    products: [
      {
        name: 'Product 1',
        price: 20,
        image:
          'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
        size: ['XS', 'S', 'M', 'L'],
        color: ['black', 'white', 'blue'],
      },
      {
        name: 'Product 2',
        price: 30,
        image:
          'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
        size: ['XS', 'S', 'M', 'L'],
        color: ['black', 'white', 'blue'],
      },
      {
        name: 'Product Detail',
        price: 20,
        image:
          'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
        size: ['XS', 'S', 'M', 'L'],
        color: ['black', 'white', 'blue'],
      },
    ],
  },
};
