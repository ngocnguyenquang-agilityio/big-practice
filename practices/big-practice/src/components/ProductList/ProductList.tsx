// Types
import { IProduct } from '@interfaces';

// Components
import ProductCard from '@components/ProductCard/ProductCard';
import { Link } from 'react-router-dom';

// Constants
import { APP_ROUTERS } from '../../constants/routes';

const ProductList = ({ products = [], sortBy }: { products: IProduct[]; sortBy?: string }) => {
  return (
    <ul
      data-testid='product-list'
      className='grid grid-flow-row gap-4 grid-cols-1 sm:grid-cols-2 lg:grid-cols-3'
    >
      {products
        .sort((a: IProduct, b: IProduct) => {
          if (sortBy === 'low-to-high') {
            return a.price > b.price ? 1 : -1;
          }
          if (sortBy === 'high-to-low') {
            return a.price < b.price ? 1 : -1;
          }

          return 0;
        })
        .map((product) => (
          <li
            className='aspect-square transition-opacity animate-fadeIn'
            key={product.id}
          >
            <Link to={`${APP_ROUTERS.DETAILS_PAGE}/${product.id}`}>
              <ProductCard
                title={product.title}
                price={product.price}
                thumbnail={product.thumbnail}
              />
            </Link>
          </li>
        ))}
    </ul>
  );
};

export default ProductList;
