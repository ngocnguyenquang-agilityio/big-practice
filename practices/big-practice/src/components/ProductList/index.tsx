// Types
import { IProduct } from '@interfaces';

// Components
import ProductCard from '@components/ProductCard';

const ProductList = ({ products = [], sortBy }: { products: IProduct[]; sortBy: string }) => {
  return (
    <ul className='grid grid-flow-row gap-4 grid-cols-1 sm:grid-cols-2 lg:grid-cols-3'>
      {products
        .sort((a, b): any => {
          if (sortBy === 'low-to-high') {
            return a.price > b.price ? 1 : -1;
          }
          if (sortBy === 'high-to-low') {
            return a.price < b.price ? 1 : -1;
          }
        })
        .map((item) => (
          <li
            className='aspect-square transition-opacity animate-fadeIn'
            key={item.id}
          >
            <ProductCard
              title={item.title}
              price={item.price}
              thumbnail={item.thumbnail}
            />
          </li>
        ))}
    </ul>
  );
};

export default ProductList;
