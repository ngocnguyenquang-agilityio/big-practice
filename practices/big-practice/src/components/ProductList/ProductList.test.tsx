import { RenderResult, render } from '@testing-library/react';

// Component
import ProductList from '.';

const mockProducts = [
  {
    id: 1,
    title: 'Product 1',
    price: 20,
    category: 'phone',
    images: [
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
    ],
    thumbnail:
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
  },
  {
    id: 2,
    title: 'Product 2',
    price: 20,
    category: 'phone',
    images: [
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
    ],
    thumbnail:
      'https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75',
  },
];

describe('ProductList Component', () => {
  let comp: RenderResult;
  beforeEach(() => {
    comp = render(
      <ProductList
        products={mockProducts}
        sortBy='low-to-high'
      />,
    );
  });

  test('Should render successfully', () => {
    expect(comp).toBeTruthy();
  });

  test('Should match snapshot', () => {
    expect(comp).toMatchSnapshot();
  });
});
