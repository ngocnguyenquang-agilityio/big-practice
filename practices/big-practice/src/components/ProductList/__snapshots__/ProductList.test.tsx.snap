// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="grid grid-flow-row gap-4 grid-cols-1 sm:grid-cols-2 lg:grid-cols-3"
      >
        <li
          class="aspect-square transition-opacity animate-fadeIn"
        >
          <div
            class="relative inline-block h-full w-full"
          >
            <div
              class="group flex h-full w-full items-center justify-center overflow-hidden rounded-lg border hover:border-blue-600 bg-black relative border-neutral-800"
            >
              <img
                alt="Product 1"
                class="relative h-full w-full object-cover transition duration-300 ease-in-out group-hover:scale-105"
                loading="lazy"
                src="https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75"
              />
              <div
                class="absolute bottom-0 left-0 flex w-full px-4 pb-4"
              >
                <div
                  class="flex items-center rounded-full border p-1 text-xs font-semibold backdrop-blur-md border-neutral-800 bg-black/70 text-white"
                >
                  <h3
                    class="mr-4 line-clamp-2 flex-grow pl-2 leading-none tracking-tight"
                  >
                    Product 1
                  </h3>
                  <p
                    class="flex-none rounded-full bg-blue-600 p-2 text-white"
                  >
                    20
                    <span
                      class="ml-1 inline @[275px]/label:inline"
                    >
                      USD
                    </span>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li
          class="aspect-square transition-opacity animate-fadeIn"
        >
          <div
            class="relative inline-block h-full w-full"
          >
            <div
              class="group flex h-full w-full items-center justify-center overflow-hidden rounded-lg border hover:border-blue-600 bg-black relative border-neutral-800"
            >
              <img
                alt="Product 2"
                class="relative h-full w-full object-cover transition duration-300 ease-in-out group-hover:scale-105"
                loading="lazy"
                src="https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75"
              />
              <div
                class="absolute bottom-0 left-0 flex w-full px-4 pb-4"
              >
                <div
                  class="flex items-center rounded-full border p-1 text-xs font-semibold backdrop-blur-md border-neutral-800 bg-black/70 text-white"
                >
                  <h3
                    class="mr-4 line-clamp-2 flex-grow pl-2 leading-none tracking-tight"
                  >
                    Product 2
                  </h3>
                  <p
                    class="flex-none rounded-full bg-blue-600 p-2 text-white"
                  >
                    20
                    <span
                      class="ml-1 inline @[275px]/label:inline"
                    >
                      USD
                    </span>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="grid grid-flow-row gap-4 grid-cols-1 sm:grid-cols-2 lg:grid-cols-3"
    >
      <li
        class="aspect-square transition-opacity animate-fadeIn"
      >
        <div
          class="relative inline-block h-full w-full"
        >
          <div
            class="group flex h-full w-full items-center justify-center overflow-hidden rounded-lg border hover:border-blue-600 bg-black relative border-neutral-800"
          >
            <img
              alt="Product 1"
              class="relative h-full w-full object-cover transition duration-300 ease-in-out group-hover:scale-105"
              loading="lazy"
              src="https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75"
            />
            <div
              class="absolute bottom-0 left-0 flex w-full px-4 pb-4"
            >
              <div
                class="flex items-center rounded-full border p-1 text-xs font-semibold backdrop-blur-md border-neutral-800 bg-black/70 text-white"
              >
                <h3
                  class="mr-4 line-clamp-2 flex-grow pl-2 leading-none tracking-tight"
                >
                  Product 1
                </h3>
                <p
                  class="flex-none rounded-full bg-blue-600 p-2 text-white"
                >
                  20
                  <span
                    class="ml-1 inline @[275px]/label:inline"
                  >
                    USD
                  </span>
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        class="aspect-square transition-opacity animate-fadeIn"
      >
        <div
          class="relative inline-block h-full w-full"
        >
          <div
            class="group flex h-full w-full items-center justify-center overflow-hidden rounded-lg border hover:border-blue-600 bg-black relative border-neutral-800"
          >
            <img
              alt="Product 2"
              class="relative h-full w-full object-cover transition duration-300 ease-in-out group-hover:scale-105"
              loading="lazy"
              src="https://demo.vercel.store/_next/image?url=https%3A%2F%2Fcdn.shopify.com%2Fs%2Ffiles%2F1%2F0754%2F3727%2F7491%2Ffiles%2Ft-shirt-1.png%3Fv%3D1689798965&w=3840&q=75"
            />
            <div
              class="absolute bottom-0 left-0 flex w-full px-4 pb-4"
            >
              <div
                class="flex items-center rounded-full border p-1 text-xs font-semibold backdrop-blur-md border-neutral-800 bg-black/70 text-white"
              >
                <h3
                  class="mr-4 line-clamp-2 flex-grow pl-2 leading-none tracking-tight"
                >
                  Product 2
                </h3>
                <p
                  class="flex-none rounded-full bg-blue-600 p-2 text-white"
                >
                  20
                  <span
                    class="ml-1 inline @[275px]/label:inline"
                  >
                    USD
                  </span>
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
